econml/tests/dgp.py

```diff
@@ -0,0 +1,166 @@
+        self.sigma_x = sigma_x
+        self.sigma_y = sigma_y
+        self.hetero_inds = hetero_inds.astype(
+            int) if hetero_inds is not None else hetero_inds
+        self.endo_inds = np.setdiff1d(
+            np.arange(self.n_x), hetero_inds).astype(int)
+        # The first s_x state variables are confounders. The final s_x variables are exogenous and can create
+        # heterogeneity
+        self.Alpha = np.random.uniform(-1, 1,
+                                       size=(self.n_x, self.n_treatments))
+        self.Alpha /= np.linalg.norm(self.Alpha, axis=1, ord=1, keepdims=True)
+        self.Alpha *= state_effect
+        if self.hetero_inds is not None:
+            self.Alpha[self.hetero_inds] = 0
+
+        self.Beta = np.zeros((self.n_x, self.n_x))
+        for t in range(self.n_x):
+            self.Beta[t, :] = autoreg * np.roll(np.random.uniform(low=4.0**(-np.arange(
+                0, self.n_x)), high=4.0**(-np.arange(1, self.n_x + 1))), t)
+        if self.hetero_inds is not None:
+            self.Beta[np.ix_(self.endo_inds, self.hetero_inds)] = 0
+            self.Beta[np.ix_(self.hetero_inds, self.endo_inds)] = 0
+
+        self.epsilon = np.random.uniform(-1, 1, size=self.n_treatments)
+        self.zeta = np.zeros(self.n_x)
+        self.zeta[:self.s_x] = self.conf_str / self.s_x
+
+        self.y_hetero_effect = np.zeros(self.n_x)
+        self.x_hetero_effect = np.zeros(self.n_x)
+        if self.hetero_inds is not None:
+            self.y_hetero_effect[self.hetero_inds] = np.random.uniform(.5 * hetero_strength, 1.5 * hetero_strength) / \
+                len(self.hetero_inds)
+            self.x_hetero_effect[self.hetero_inds] = np.random.uniform(.5 * hetero_strength, 1.5 * hetero_strength) / \
+                len(self.hetero_inds)
+
+        self.true_effect = np.zeros((self.n_periods, self.n_treatments))
+        # Invert indices to match latest API
+        self.true_effect[self.n_periods - 1] = self.epsilon
+        for t in np.arange(self.n_periods - 2, -1, -1):
+            self.true_effect[t, :] = (self.zeta.reshape(
+                1, -1) @ np.linalg.matrix_power(self.Beta, (self.n_periods - 1 - t) - 1) @ self.Alpha)
+
+        self.true_hetero_effect = np.zeros(
+            (self.n_periods, (self.n_x + 1) * self.n_treatments))
+        self.true_hetero_effect[self.n_periods - 1, :] = cross_product(
+            add_constant(self.y_hetero_effect.reshape(1, -1), has_constant='add'),
+            self.epsilon.reshape(1, -1))
+        for t in np.arange(self.n_periods - 2, -1, -1):
+            # Invert indices to match latest API
+            self.true_hetero_effect[t, :] = cross_product(
+                add_constant(self.x_hetero_effect.reshape(1, -1), has_constant='add'),
+                self.zeta.reshape(1, -1) @ np.linalg.matrix_power(
+                    self.Beta, (self.n_periods - 1 - t) - 1) @ self.Alpha)
+        return self
+
+    def hetero_effect_fn(self, t, x):
```

Comment:
I think the dgp still allows heterogneeity with respect to the time -varying X, which will be incompatible wiht what the method asssumes, which is only wrt X[0].

Reply:
my bad, you're doing this correctly